modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/IndexScanNode.java

```diff
@@ -233,80 +247,65 @@ private void requestNextBatch() {
-
-                if (rangeConditions == null) {
-                    flags = SortedIndex.INCLUDE_LEFT | SortedIndex.INCLUDE_RIGHT;
-                    curPartIdx++;
-                } else {
-                    if (rangeConditionIterator == null) {
-                        rangeConditionIterator = rangeConditions.iterator();
-                    }
-
-                    RangeCondition<RowT> cond = rangeConditionIterator.next();
+        } else if (!rangeConditionsProcessed) {
+            RangeCondition<RowT> cond = null;
 
-                    lowerBound = toBinaryTuplePrefix(cond.lower());
-                    upperBound = toBinaryTuplePrefix(cond.upper());
+            if (rangeConditionIterator == null || !rangeConditionIterator.hasNext()) {
+                rangeConditionsProcessed = true;
+            } else {
+                cond = rangeConditionIterator.next();
 
-                    flags |= (cond.lowerInclude()) ? SortedIndex.INCLUDE_LEFT : 0;
-                    flags |= (cond.upperInclude()) ? SortedIndex.INCLUDE_RIGHT : 0;
+                rangeConditionsProcessed = !rangeConditionIterator.hasNext();
+            }
 
-                    if (!rangeConditionIterator.hasNext()) { // Switch to next partition and reset range index.
-                        rangeConditionIterator = null;
-                        curPartIdx++;
-                    }
-                }
+            List<Flow.Publisher<BinaryTuple>> partPublishers = new ArrayList<>(parts.length);
 
-                ((SortedIndex) schemaIndex.index()).scan(
-                        parts[part],
-                        context().transaction(),
-                        lowerBound,
-                        upperBound,
-                        flags,
-                        requiredColumns
-                ).subscribe(new SubscriberImpl());
-            } else {
-                assert schemaIndex.type() == Type.HASH;
+            for (int p : parts) {
+                partPublishers.add(partPublisher(p, cond));
+            }
 
-                int part = curPartIdx;
-                BinaryTuple key = null;
+            SubscriberImpl subscriber = new SubscriberImpl();
 
-                if (rangeConditions == null) {
-                    curPartIdx++;
-                } else {
-                    if (rangeConditionIterator == null) {
-                        rangeConditionIterator = rangeConditions.iterator();
-                    }
+            CompositeSubscription<BinaryTuple> compSubscription = comp != null
+                    ? new OrderedMergeCompositeSubscription<>(subscriber, comp, inBufSize, partPublishers.size())
+                    : new CompositeSubscription<>(subscriber);
 
-                    RangeCondition<RowT> cond = rangeConditionIterator.next();
+            compSubscription.subscribe(partPublishers);
 
-                    assert cond.lower() == cond.upper();
+            subscriber.onSubscribe(compSubscription);
```

Comment:
To be honest, I don't get this part. Why does ScanNode invoke `onSubscribe` method? Why do publishers are subscribing to subscription?

The flow should be like this:
```
indexPublisher(<some args related to an index scan>).subscribe(new SubscriberImpl())
```
, where `indexPublisher()` returns a Publisher<RowT>

And that's all the ScanNode should be aware of

Reply:
Done